llama : simplify Mamba with advanced batch splits

compilade opened the pull request:
As promised in <https://github.com/ggerganov/llama.cpp/pull/7531#issuecomment-2209585346>, I've been extracting the advanced batch splits out of the Jamba PR (#7531).

I've also backported the contiguous allocation of recurrent state slots, which makes it possible to *also* include the changes from #7531 which simplify the `ggml` operators used specifically for Mamba. Hopefully this isn't too much at once.

See <https://github.com/ggerganov/llama.cpp/pull/7531#discussion_r1620997020> for an explanation of the batch splits.

# Summary

- `ggml.c`
  - Simplify `ggml_ssm_conv` and `ggml_ssm_scan` by assuming batched sequences have the same number of new tokens, and that the states are contiguous and ordered correctly.
  - Allow `ggml_concat` to work with a non-contiguous second argument.
    - The CPU implementation already supported this, but it was guarded with an assertion. Meanwhile, I *think* the CUDA implementation already supports this too, and does *not* prevent its usage (not totally sure), so I did not change it.
- `llama.cpp`
  - Advanced batch splits handled with `lctx.sbatch` for persistent buffers
    - Refactor "helpers for smoother batch API transition", by handling them in `llama_sbatch`, which allows avoiding repeated allocations by re-using the same buffers.
    - Simple batch splits should be equivalent to the previous behavior and are made with `lctx.sbatch.split_simple(n_tokens)` to build a `llama_ubatch` with a max size of `n_tokens`.
    - Equal-sequence-lengths splits are made with `lctx.sbatch.split_equal(n_tokens)`, and are used to simplify the operators of recurrent models.
    - Add `llama_ubatch`. Similar to `llama_batch`, but aware of equal-length sequences.
      - Make `llama_set_inputs` (and others) use `llama_ubatch` instead of `llama_batch`.
  - Make recurrent state slot allocation contiguous in `llama_kv_cache_find_slot`
  - Add `llm_build_mamba` to build a Mamba block, used for Mamba, and will be used for Jamba
  - Add `llm_build_copy_mask_state` (maybe not a good name) to abstract away the shuffling and masking of recurrent states. Used for Mamba, and it should be usable for other recurrent architectures too.
  - Simplify the sanity checks for `qs.n_attention_wv` in `llama_model_quantize_internal` to make it future proof for hybrid models.
  - Reorder the outputs when using advanced batch splits like `split_equal` in conjunction with `llama_get_logits`, because the API makes it so that the outputs should have the same order they had as the user-provided batch, not something based on batch split rules.

For simplicity, this does ***not*** include the separation of the KV cache and the recurrent state cache. Both still use the same buffers (`lctx.kv_self.k_l`, and `lctx.kv_self.v_l`, as on `master`). The separation (necessary for hybrid models) will be introduced at the same time as Jamba.

# TODO

- [x] Test the slot allocation of `llama_kv_cache_find_slot` with the `--hellaswag` benchmark in `llama-perplexity` with a Mamba model
  - This uses lots of parallel sequences in an unusual way, and so I think it's a great stress test.
- [x] Session file saving and reloading
  - [x] Reloading needs to rebuild the `tail` metadata for recurrent states. (i.e. which cell is the end of which sequence)
  - [x] The server tests need to pass
- [x] Make sure T5 still works
- [x] Make sure the pooled embeddings still work
  - tested `bge-small` with `llama-embeddings` with parallel prompts with `--pooling cls`, `--pooling last` and `--pooling mean`; results exactly match `master`. 
- [x] Make sure Gemma's sliding window mask still works
- [x] Decide whether to rename `llama_reorder_outputs` to `llama_output_reorder` and move it close to `llama_output_reserve`.
  - renamed and moved

# Future ideas

- whole-sequence splits for embeddings
- handle pooling types like `cls` and `last` within the `ubatch.outputs` when splitting a batch; `inp_cls` is redundant with `inp_out_ids`.

---

- [x] I have read the [contributing guidelines](https://github.com/ggerganov/llama.cpp/blob/master/CONTRIBUTING.md)
- Self-reported review complexity:
  - [x] Medium

compilade committed:
llama : advanced batch splits

This includes equal-sequence-length batch splits which are useful
to simplify recurrent model operators.

* llama : always make recurrent state slots contiguous

* ggml : simplify mamba operators

compilade committed:
Merge branch 'master' into compilade/batch-splits

compilade committed:
llama : fix integer signedness mixing

compilade committed:
llama : logits_all has priority over batch->logits

Otherwise, the server embeddings tests failed.
This was likely an existing problem but was only detected here
because of an additional assertion.

ggerganov commented on src/llama.cpp:
```diff
@@ -2732,6 +2750,341 @@ struct llama_model {
+            return;
+        }
+        std::sort(ids.begin(), ids.end(),
+            [batch](size_t a, size_t b) {
```
```suggestion
            [&batch](size_t a, size_t b) {
```

ggerganov commented on src/llama.cpp:
```diff
@@ -2732,6 +2750,341 @@ struct llama_model {
+        llama_sbatch_seq * last_seq = nullptr;
+
+        if (batch.n_seq_id != nullptr && batch.seq_id != nullptr) {
+            for (size_t i = 0; i < n_tokens; ++i) {
+                const size_t bi = ids[i];
+                const size_t s_len = seq.size();
+                const int32_t n_seqs = batch.n_seq_id[bi];
+                llama_seq_id * seq_ids = batch.seq_id[bi];
+                if (last_seq != nullptr) {
+                    bool same = n_seqs == last_seq->n_seq_id;
+                    for (int32_t j = 0; same && j < n_seqs; ++j) {
+                        if (seq_ids[j] != last_seq->seq_id[j]) {
+                            same = false;
+                        }
+                    }
+                    if (same) {
+                        last_seq->length += 1;
+                        continue;
+                    }
+                }
+                llama_sbatch_seq new_seq = {n_seqs, seq_ids, i, 1, batch.all_seq_id};
+                seq.push_back(new_seq);
+                last_seq = &seq[s_len];
+            }
```
```diff
@@ -3050,7 +3051,6 @@ struct llama_sbatch {
         if (batch.n_seq_id != nullptr && batch.seq_id != nullptr) {
             for (size_t i = 0; i < n_tokens; ++i) {
                 const size_t bi = ids[i];
-                const size_t s_len = seq.size();
                 const int32_t n_seqs = batch.n_seq_id[bi];
                 llama_seq_id * seq_ids = batch.seq_id[bi];
                 if (last_seq != nullptr) {
@@ -3067,7 +3067,7 @@ struct llama_sbatch {
                 }
                 llama_sbatch_seq new_seq = {n_seqs, seq_ids, i, 1, batch.all_seq_id};
                 seq.push_back(new_seq);
-                last_seq = &seq[s_len];
+                last_seq = &seq.back();
             }
         } else {
             llama_sbatch_seq new_seq = {1, nullptr, 0, n_tokens, batch.all_seq_id};
```

ggerganov commented on src/llama.cpp:
```diff
@@ -14530,55 +15077,31 @@ static int llama_decode_internal(
-        };
+    while (lctx.sbatch.n_tokens > 0) {
+        // For now, only use equal splits for recurrent model architectures
+        llama_ubatch u_batch = kv_self.recurrent ? lctx.sbatch.split_equal(n_ubatch) : lctx.sbatch.split_simple(n_ubatch);
```
```suggestion
        llama_ubatch ubatch = kv_self.recurrent ? lctx.sbatch.split_equal(n_ubatch) : lctx.sbatch.split_simple(n_ubatch);
```

ggerganov reviewed (commented):
Tested `t5-small` and it currently segfaults - let me know if you need help with resolving it

compilade committed:
llama : apply suggestions

Co-authored-by: Georgi Gerganov <ggerganov@gmail.com>

compilade committed:
llama : fix t5 segfault

ggerganov commented:
> Make sure Gemma's sliding window mask still works

The following command produces identical perplexity on `master` and this branch:

```bash
./llama-perplexity \
    -m models/gemma-2-9b/ggml-model-f16.gguf \
    -f build/wikitext-2-raw/wiki.test.raw \
    -ngl 99 -c 8192
```

Is this enough to confirm the SWA functionality?

compilade commented:
> Is this enough to confirm the SWA functionality?

I think so. Might also be relevant to test SWA with parallel sequences too (I *think* this is what using a bigger `-b` (and `-ub`?) than `-c` does with `llama-perplexity`).

hackey commented:
Guys, is there any progress in supporting Mamba2 (I'm interested in the new mamba-codestral)?

compilade committed:
Merge branch 'master' into compilade/batch-splits

compilade commented:
> Guys, is there any progress in supporting Mamba2 (I'm interested in the new mamba-codestral)?

Still waiting on some upstream changes (see <https://huggingface.co/mistralai/mamba-codestral-7B-v0.1/discussions/1>), but otherwise I'm beginning to investigate the conversion for Mamba2 models, at least to have some GGUFs (even with no inference support) to experiment with implementing it.

First thing I'm noticing is the lack of metadata in the `config.json` of Mamba2 models. No state size, no convolution kernel size, no time step rank, and in the case of `mamba-codestral-7B-v0.1`, no indication that it's a Mamba2 model, except from the tensor names and sizes.
For the state sizes, I guess these are hardcoded in the `state-spaces/mamba` implementation, in which case I'll hardcode them here too and/or find what is used to calculate them.

I've also recently started to simplify the session file save & restore code in `llama.cpp` (but I'll likely open a separate PR, since I think that refactor is best tested on its own), because I'm noticing that it's often causing me problems to adapt it to changes to the KV cache structure, due to there being at least 4 places needing to be updated and/or considered for each change (read/write + seq read/write). So I'll be unifying these code paths to make them easier to maintain.

hackey commented:
> > Guys, is there any progress in supporting Mamba2 (I'm interested in the new mamba-codestral)?
> 
> Still waiting on some upstream changes (see https://huggingface.co/mistralai/mamba-codestral-7B-v0.1/discussions/1), but otherwise I'm beginning to investigate the conversion for Mamba2 models, at least to have some GGUFs (even with no inference support) to experiment with implementing it.
> 
> First thing I'm noticing is the lack of metadata in the `config.json` of Mamba2 models. No state size, no convolution kernel size, no time step rank, and in the case of `mamba-codestral-7B-v0.1`, no indication that it's a Mamba2 model, except from the tensor names and sizes. For the state sizes, I guess these are hardcoded in the `state-spaces/mamba` implementation, in which case I'll hardcode them too and/or find what is used to calculate them.
> 
> I've also recently started to simplify the session file save & restore code in `llama.cpp` (but I'll likely open a separate PR, since I think that refactor is best tested on its own), because I'm noticing that it's often causing me problems to adapt it to changes to the KV cache structure, due to there being at least 4 places needing to be updated and/or considered for each change (read/write + seq read/write). So I'll be unifying these code paths to make them easier to maintain.

I also encountered difficulties running mamba-codestral. I tried to run this model on https://github.com/state-spaces/mamba. But there is no config.json in the model repository. mamba-codestral includes a new tokenizer v3.
Although Mistral writes that the model can be run on state-spaces/mamba, nothing worked for me.

Please see the discussion here:
https://github.com/NVIDIA/TensorRT-LLM/issues/1968
and a few hours ago an example for running mamba appeared:
https://github.com/NVIDIA/TensorRT-LLM/tree/main/examples/mamba

Maybe this will help development.

ggerganov commented:
> For the state sizes, I guess these are hardcoded in the state-spaces/mamba implementation, in which case I'll hardcode them too and/or find what is used to calculate them.

Yes, we can hardcode initially

> I've also recently started to simplify the session file save & restore code in llama.cpp (but I'll likely open a separate PR, since I think that refactor is best tested on its own), because I'm noticing that it's often causing me problems to adapt it to changes to the KV cache structure, due to there being at least 4 places needing to be updated and/or considered for each change (read/write + seq read/write). So I'll be unifying these code paths to make them easier to maintain.

Sounds good - a separate PR would be easier to review

Regarding Codestral - want to highlight again the comment by Mistral team about `ngroups = 8`: https://github.com/ggerganov/llama.cpp/issues/8519#issuecomment-2233110921. Seems important

compilade committed:
Merge branch 'master' into compilade/batch-splits

compilade committed:
llama : fix Mamba session save and restore

awgr commented:
> > Guys, is there any progress in supporting Mamba2 (I'm interested in the new mamba-codestral)?
> 
> Still waiting on some upstream changes (see https://huggingface.co/mistralai/mamba-codestral-7B-v0.1/discussions/1), but otherwise I'm beginning to investigate the conversion for Mamba2 models, at least to have some GGUFs (even with no inference support) to experiment with implementing it.
> 
> First thing I'm noticing is the lack of metadata in the `config.json` of Mamba2 models. No state size, no convolution kernel size, no time step rank, and in the case of `mamba-codestral-7B-v0.1`, no indication that it's a Mamba2 model, except from the tensor names and sizes. For the state sizes, I guess these are hardcoded in the `state-spaces/mamba` implementation, in which case I'll hardcode them too and/or find what is used to calculate them.
> 
> I've also recently started to simplify the session file save & restore code in `llama.cpp` (but I'll likely open a separate PR, since I think that refactor is best tested on its own), because I'm noticing that it's often causing me problems to adapt it to changes to the KV cache structure, due to there being at least 4 places needing to be updated and/or considered for each change (read/write + seq read/write). So I'll be unifying these code paths to make them easier to maintain.

https://github.com/state-spaces/mamba/blob/main/mamba_ssm/modules/mamba2_simple.py

This includes some details that may be interesting for you.

compilade committed:
llama : minor cosmetic changes

compilade committed:
Merge branch 'master' into compilade/batch-splits

compilade committed:
llama : rename llama_reorder_outputs to llama_output_reorder

Also move it closer to llama_output_reserve.

* llama : fix pooled embeddings when using batches with equal_seqs

ggerganov committed:
minor : add struct members for clarity

ggml-ci

ggerganov reviewed (approved):
In the future we should refactor the KV cache using object-oriented design so that the implementation of non-recurrent, recurrent and other modes are better separated and easier to read.

compilade committed:
Merge branch 'master' into compilade/batch-splits

compilade commented:
I'll be re-running a few tests before merging this in hopefully less than 2 days. There is now both [Mamba-2](https://github.com/ggerganov/llama.cpp/issues/8519#issuecomment-2295593409) and [RWKV v6](https://github.com/ggerganov/llama.cpp/pull/8980) which kind of need this to simplify the implementation.

compilade committed:
llama : fix T5 segfault again

compilade commented:
I've ran some tests, and there's a problem: pooled embeddings with Mamba can't work with multiple sequences anymore.

This is because `lctx.embd_seq` is overwritten at each `ubatch` which makes it only work if everything fits in a single `ubatch`, which is not the case when sequences don't have the same length and are split to makes them all equal in Mamba's `ubatch`.

This could be fixed by letting causal embeddings be split over multiple `ubatch`. I'll try to find a way to do this cleanly.

- [x] perplexity
  - [x] v0-mamba-100k
    - [x] 1 chunk per batch
    - [x] 4 chunks per batch
    - [x] 4 batches per chunk
    - [x] 4 ubatches per batch
  - [x] v0-llama2-100k
    - [x] 1 chunk per batch
    - [x] 4 chunks per batch
    - [x] 4 batches per chunk
    - [x] 4 ubatches per batch
- [ ] `llama-embedding`
  - [x] t5-small
    - fails with `GGML_ASSERT(strcmp(embd->name, "result_norm") == 0 && "missing result_output tensor")` as on `master`.
  - [ ] mamba-130M
    - ❌ Does NOT work with more than one sequence anymore 
  - [x] bge-small
- [x] `llama-parallel -c 1024 -np 5 -ns 7 --seed 42 --temp 1`
  - [x] t5-small
    - fails to decode, as on `master` (fixed segfault in <https://github.com/ggerganov/llama.cpp/pull/8526/commits/652e9b0d6145a3e97d14e2a10efa531f1d76dc31>)
  - [x] stories-MoE
    - works with `-c 1024`, but segfaults otherwise (as on `master`)
  - [x] mamba-130M
    - works without problem
- [x] `perplexity --hellaswag` (parallel sequences of uneven length, also the only test with batches having more than one seq_id per token)
  - [x] v0-mamba-100k
  - [x] v0-llama2-100k
- [x] Gemma2 sliding window
  - [x] Gemma2-2B perplexity with `-c 5120` (its sliding window is 4096), first chunk generates the same perplexity.
  - [ ] parallel sliding windows (on more than one sequence)
    - not sure how to test that

compilade committed:
llama : fix Mamba pooled embeddings with multiple sequences

Until the pooled embeddings are refactored to allow splitting
across ubatches for causal embeddings,
recurrent models can only process a single sequence per ubatch
when calculating pooled embeddings.

compilade committed:
llama : add llama_model_is_recurrent to simplify figuring that out

This will make it easier to more cleanly support RWKV-v6 and Mamba-2.

compilade commented:
I've fixed the pooled embeddings problem with Mamba in <https://github.com/ggerganov/llama.cpp/pull/8526/commits/b264eddbb26c695d50d04c37a5b9bb91181bc551> by making it only process a single sequence per `ubatch`. When the sequences are short, this is slightly slower than processing them all at once, unfortunately.

In the future, the pooled embeddings will be refactored to allow causal embeddings to be split across ubatches. It should also be possible to remove `inp_cls`, because it's redundant with `inp_out_ids`. `LLAMA_POOLING_TYPE_CLS` and `LLAMA_POOLING_TYPE_LAST` could be handled directly when splitting batches, since they only affect which tokens get their output selected. `LLAMA_POOLING_TYPE_MEAN` will be a bit harder to allow splitting, but since the total number of tokens per sequence per batch is known in advance, there might still be a way.

I'm postponing that pooled embeddings refactor to another PR. I consider this ready.

compilade committed:
Merge branch 'master' into compilade/batch-splits

ggerganov reviewed (approved):
Great work as always!

Let's merge and resolve any remaining issues from `master`. I'll follow up with the initial SSM Metal kernels shortly after (#8546)

compilade committed:
llama : fix simple splits when the batch contains embeddings

awgr commented:
@compilade btw, I have the SSD implementation on CPU, more or less, if it's
interesting for you.

On Wed, Aug 21, 2024 at 2:58 PM compilade ***@***.***> wrote:

> Merged #8526 <https://github.com/ggerganov/llama.cpp/pull/8526> into
> master.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ggerganov/llama.cpp/pull/8526#event-13967053247>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BGA7QVODOGJ22XVGJENS4D3ZSUEQ5AVCNFSM6AAAAABK7U7OGKVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJTHE3DOMBVGMZDINY>
> .
> You are receiving this because you commented.Message ID: <ggerganov/llama.
> ***@***.***>
>

mann1x commented:
@compilade 

I get this error quantizing deepseek2 since the merge of this PR:
https://github.com/ggerganov/llama.cpp/issues/9155
